Framework/MDAlgorithms/src/SmoothMD.cpp

```diff
@@ -186,6 +281,126 @@ SmoothMD::hatSmooth(IMDHistoWorkspace_const_sptr toSmooth,
+  }
+
+  const int nThreads = Mantid::API::FrameworkManager::Instance()
+                           .getNumOMPThreads(); // NThreads to Request
+
+  auto read_ws = outWS;
+  auto write_ws = tempWS;
+  for (size_t dimension_number = 0; dimension_number < widthVector.size();
+       ++dimension_number) {
+
+    auto iterators = toSmooth->createIterators(nThreads, NULL);
+
+    // Alternately write to each workspace
+    if (dimension_number % 2 == 0) {
+      read_ws = outWS;
+      write_ws = tempWS;
+    } else {
+      read_ws = tempWS;
+      write_ws = outWS;
+    }
+
+    PARALLEL_FOR_NO_WSP_CHECK()
+    for (int it = 0; it < int(iterators.size()); ++it) {
+
+      PARALLEL_START_INTERUPT_REGION
+      boost::scoped_ptr<MDHistoWorkspaceIterator> iterator(
+          dynamic_cast<MDHistoWorkspaceIterator *>(iterators[it]));
+
+      if (!iterator) {
+        throw std::logic_error(
+            "Failed to cast IMDIterator to MDHistoWorkspaceIterator");
+      }
+
+      do {
+
+        // Gets linear index at current position
+        size_t iteratorIndex = iterator->getLinearIndex();
+
+        if (useWeights) {
+
+          // Check that we could measure here.
+          if ((*weightingWS)->getSignalAt(iteratorIndex) == 0) {
+
+            write_ws->setSignalAt(iteratorIndex,
+                                  std::numeric_limits<double>::quiet_NaN());
+            write_ws->setErrorSquaredAt(
+                iteratorIndex, std::numeric_limits<double>::quiet_NaN());
+
+            continue; // Skip we couldn't measure here.
+          }
+        }
+
+        std::pair<std::vector<size_t>, std::vector<bool>> indexesAndValidity =
+            iterator->findNeighbourIndexesByWidth1D(
+                static_cast<int>(gaussian_kernels[dimension_number].size()),
+                static_cast<int>(dimension_number));
+        std::vector<size_t> neighbourIndexes = std::get<0>(indexesAndValidity);
+        std::vector<bool> indexValidity = std::get<1>(indexesAndValidity);
+        auto normalised_kernel = renormaliseKernel(
+            gaussian_kernels[dimension_number], indexValidity);
+
+        // Convolve signal with kernel
+        double sumSignal = 0;
+        double sumSquareError = 0;
+        double error = 0;
+        for (size_t i = 0; i < neighbourIndexes.size(); ++i) {
+          if (indexValidity[i]) {
+            sumSignal += read_ws->getSignalAt(neighbourIndexes[i]) *
+                         normalised_kernel[i];
+            error =
+                read_ws->getErrorAt(neighbourIndexes[i]) * normalised_kernel[i];
+            sumSquareError += error * error;
+          }
+
+          write_ws->setSignalAt(iteratorIndex, sumSignal);
+          write_ws->setErrorSquaredAt(iteratorIndex, sumSquareError);
+        }
+        progress.report();
+
+      } while (iterator->next());
+      PARALLEL_END_INTERUPT_REGION
+    }
+    PARALLEL_CHECK_INTERUPT_REGION
+  }
```

Comment:
seems like a very long method. should we have anonymous functions which can do parts of the work and make it more readable?